### ISSUE TITLE
Two admin users on new Rancher v2.5.6 install

### ISSUE THREAD

**Issue by dnoland1** [rancher/rancher#31609]
**What kind of request is this (question/bug/enhancement/feature request):**
bug

**Steps to reproduce (least amount of steps as possible):**
1. Install Rancher v2.5.6 HA on k3s 1.19.8 cluster with two server nodes

**Result:**
Two admin users were created:
```
NAME           USERNAME   DISPLAY_NAME                       PRINCIPAL_IDS                           CREATE_DATE
u-b4qkhsnliz   <none>     System account for Cluster local   [system://local local://u-b4qkhsnliz]   2021-03-06T03:58:03Z
user-7sphl     admin      Default Admin                      [local://user-7sphl]                    2021-03-06T03:57:25Z
user-tbftx     admin      Default Admin                      [local://user-tbftx]                    2021-03-06T03:57:26Z
```
Admin cannot log into UI:
![image](https://user-images.githubusercontent.com/468462/110196476-a431fc00-7df9-11eb-9647-dc1010a48191.png)

Seeing many errors in logs like:
```
2021/03/06 05:17:00 [ERROR] API error response 500 for POST /v3-public/localProviders/local?action=login. Cause: found more than one users with username admin
```

**Other details that may be helpful:**
This was a fresh install. Full rancher logs available on request.

**Environment information**
- Rancher version (`rancher/rancher`/`rancher/server` image tag or shown bottom left in the UI):
v2.5.6
- Installation option (single install/HA):
HA

**Cluster information**
- Cluster type (Hosted/Infrastructure Provider/Custom/Imported):
- Machine type (cloud/VM/metal) and specifications (CPU/memory):
AWS EC2 m5a.large (2 vCPU, 8GB RAM)
- Kubernetes version (use `kubectl version`):

```
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.4", GitCommit:"e87da0bd6e03ec3fea7933c4b5263d151aafd07c", GitTreeState:"clean", BuildDate:"2021-02-21T20:21:49Z", GoVersion:"go1.15.8", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.8+k3s1", GitCommit:"95fc76b206b1455db7d50e23fb45c98d6ea97dde", GitTreeState:"clean", BuildDate:"2021-02-22T19:46:43Z", GoVersion:"go1.15.8", Compiler:"gc", Platform:"linux/amd64"}
```

- Docker version (use `docker version`):
container

gz#15595

**Comment by slickwarren** [comment]
I was not able to reproduce this on 2.5.6 with a k3s cluster with 2 server nodes.  to bring up the cluster, I did the following via our automation scripts:
* created a mysql db to use for storage
* created 2 server k3s nodes [following the documentation](https://rancher.com/docs/k3s/latest/en/installation/ha/)
* configured load balancers and routing following the above documentation
* installed rancher v2.5.6 using helm
I was able to successfully login to the cluster with both nodes healthy on version `v1.19.8+k3s1`
<img width="1086" alt="Screen Shot 2021-03-08 at 2 23 20 PM" src="https://user-images.githubusercontent.com/16691014/110389858-f94c5880-8019-11eb-82f7-3b65331752c3.png">

**Comment by sowmyav27** [comment]
@slickwarren Can you try it with the local cluster created using latest RKE ?

**Comment by slickwarren** [comment]
I also tried using the latest RKE (v1.2.6), and still could not reproduce. We think this is some sort of race condition that would rarely happen.

**Comment by samjustus** [comment]
have we seen this issue again since creation?

**Comment by dnoland1** [comment]
I've only seen it once myself.

**Comment by maxsokolovsky** [comment]
I also have not been able to reproduce the behavior using a K3s `v1.21.9+k3s1` HA with two server nodes and Postgres 11 datastore. Rancher `v2.5.12` deployed via Helm. I only had one user, who was the Default Admin. Set the password and authenticated successfully.

A couple of questions, @dnoland1:
1. Was the datastore backing the two server nodes external? If so, what datastore was used in your K3s cluster?
2. What API object are you querying in your example above - is this `user`?

**Comment by dnoland1** [comment]
@maxsokolovsky 

>    Was the datastore backing the two server nodes external? If so, what datastore was used in your K3s cluster?

Yes, MySQL.

>    What API object are you querying in your example above - is this user?

Yes, users.management.cattle.io

**Comment by MbolotSuse** [comment]
My guess on this issue is the following:
- The admin user can be created in two places - In the [Add](https://github.com/rancher/rancher/blob/58bfcb19605d51bf24f0aa5bf1d60a68b1ae2a27/pkg/data/dashboard/add.go#L18) part of the dashboard package, and in [Role_Data](https://github.com/rancher/rancher/blob/58bfcb19605d51bf24f0aa5bf1d60a68b1ae2a27/pkg/data/management/role_data.go#L447)
- In Rancher's current versions (linked above) the code that registers the admin in the Add part of the dashboard package is only called when the MCMAgent Flag is not enabled (i.e. if you are in the local cluster). 
- In v2.5.6, [Add](https://github.com/rancher/rancher/blob/65f7c844267bf7336a38ee6ea3e0e63af9e21274/pkg/data/dashboard/add.go#L17), does not have this limitation. 

It's possible that the Add + RoleData parts were called simultaneously, resulting in two admin users being created at the same time. If this is what happened, then I would guess that this was fixed long ago (probably by #32310). Given that no-one was able to reproduce it at that time, and that we have a plausible explanation that suggests the issue is fixed, I'd recommend that this issue be closed. If we see new user reports of the issue (or can repro it on a newer version), then we can re-open it and go from there. @samjustus

**Comment by MbolotSuse** [comment]
I'm going to close this as already fixed. The [Bootstrap Admin function in 2.5.6](https://github.com/rancher/rancher/blob/65f7c844267bf7336a38ee6ea3e0e63af9e21274/pkg/data/management/role_data.go#L414) did not use a mutex, and so could be subject to race conditions. The [same function in 2.8](https://github.com/rancher/rancher/blob/dda71fafe0d3bdb4f9771120dfbd42408646db35/pkg/data/management/role_data.go#L491) uses a mutex.

This appears to have been [added for the 2.6.0 release](https://github.com/rancher/rancher/pull/32334). While it's possible that there was another source of this issue, I was not able reproduce it (even on 2.5.6), and don't see any indication that it is still present today. @dnoland1 if you have continued to see this issue on a version >= 2.6.0, please let me know and we can investigate further.
